Update SYCL README (#2031)
alpaka-group/alpaka: README_SYCL.md

### General CMake options

* `alpaka_ACC_SYCL_ENABLE`: set to `ON` to enable the SYCL back-end. Requires the activation of at least one oneAPI hardware target (see below).
* `MKL_ROOT`: always required. Set to the root of your oneMKL installation. Example: `/opt/intel/oneapi/mkl/2023.1.0`.
* `oneDPL_DIR`: always required. Set to the CMake path of your oneDPL installation. Example: `/opt/intel/oneapi/dpl/2022.1.0/lib/cmake/oneDPL`.

> ### Building for CPUs

Reviewer: ```suggestion
### Building for x86 64-bit CPUs
```

The following CMake flags can be set for CPUs:

> * `alpaka_SYCL_ONEAPI_CPU`: set to `ON` to enable compilation for Intel CPUs.

Reviewer: ```suggestion
* `alpaka_SYCL_ONEAPI_CPU`: set to `ON` to enable compilation for CPUs. Relies on the Intel OpenCL CPU runtime and ahead-of-time compiler, which support Intel and AMD CPUs.
```

### Building for Intel GPUs

* `alpaka_SYCL_ONEAPI_GPU`: set to `ON` to enable compilation for Intel GPUs.

> * `alpaka_SYCL_ONEAPI_GPU_DEVICES`: semicolon-separated list of one or more Intel GPUs to compile for. The possible values for the devices are listed in the [UsersManual](https://intel.github.io/llvm-docs/UsersManual.html#generic-options) under the flag `-fsycl-targets`. Default: `intel_gpu_pvc`.

Author: Maybe out of the scope of this PR, but it is not completely clear to me why we are asking for a semicolon-separated list of devices and then change them into comma in `cmake/alpakaCommon.cmake` instead of asking directly for a comma-separated list.

Reviewer 2: Convention basically. When CMake expects a list for its native facilities it is semicolon-separated. I wanted this to behave the same way.

>   Note: currently only one target at a time can be specified (limitation of the Intel Compiler)

Reviewer 2: Hm, this seems like a downgrade from the previous mechanism. IIRC it was possible to specify multiple GPU targets in one pass before the change.

Author: yes, but with the old targets the warp size stuff does not work because the device-specific macros are not defined. That's why we moved to the new targets.

Reviewer: From some tests it's not clear if multiple targets are supported or not... still looking into it.

Reviewer: OK, after running some more tests, I think the code _is_ compiled for multiple device of the same backend, but _only the first one_ is actually used at runtime.

  ```
  See [Intel's FAQ](https://github.com/intel/compute-runtime/blob/master/opencl/doc/FAQ.md#feature-double-precision-emulation-fp64) for more information.
* The FPGA back-end does not support atomics. alpaka will not check this.

> * device global variables are not supported in SYCL yet.

Reviewer: ```suggestion
* Device global variables (corresponding to `__device__` and `__constant__` variables in CUDA) are not supported in the SYCL back-end yet.
```

### Choosing the sub-group size (warp size)

> There is a trait to specify at compile time the sub-group size to use for a kernel. The default behaviour, when no sub-group size is specified, is to let the back-end compiler pick the preferred size.

Reviewer: ```suggestion
Most SYCL targets support multiple sub-group sizes. There is a trait to specify at compile time the sub-group size to use for a kernel. The default behaviour, when no sub-group size is specified, is to let the back-end compiler pick the preferred size.
```

  ```
  See [Intel's FAQ](https://github.com/intel/compute-runtime/blob/master/opencl/doc/FAQ.md#feature-double-precision-emulation-fp64) for more information.
* The FPGA back-end does not support atomics. alpaka will not check this.
* Device global variables (corresponding to `__device__` and `__constant__` variables in CUDA) are not supported in the SYCL back-end yet.

> * The shared memory on CPU back-end does not work properly yet. The tests that use it (`atomicTest`, `blockSharedTest` and `blockSharedSharingTest`) fail with a `PI_ERROR_OUT_OF_RESOURCES`.

Reviewer 2: ```suggestion
* Shared memory does not yet work properly on CPU back-ends. The tests that use it (`atomicTest`, `blockSharedTest` and `blockSharedSharingTest`) fail with a `PI_ERROR_OUT_OF_RESOURCES`.
```

### Choosing the sub-group size (warp size)

> Most SYCL targets support multiple sub-group sizes. There is a trait to specify at compile time the sub-group size to use for a kernel. The default behaviour, when no sub-group size is specified, is to let the back-end compiler pick the preferred size.

Reviewer 2: Shouldn't we mention the trait here?

Before launching a kernel with a compile-time sub-group size the user should query the sizes supported by the device, and choose accordingly. If the device does not support the requested size, the SYCL runtime will throw a synchronous exception.

During just-in-time (JIT) compilation this guarantees that a kernel is compiled only for the sizes supported by the device. During ahead-of-time (AOT) compilation this is not enough, because the device is not known at compile time. The SYCL specification mandates that the back-end compilers should not fail if a kernel uses unsupported features, like unsupported sub-group sizes. Unfortunately the Intel OpenCL CPU and GPU compilers currently fail with a hard error. To work around this limitation, use the preprocessor macros defined when compiling AOT for the new SYCL targets to enable the compilation only for the sub-group sizes supported by each device.

> Note: while the CPU OpenCL back-end does support a sub-group size of 64, the SYCL code currently does not. To avoid issues with the sub-group primitives always consider the sub-group size of 64 as not supported by the device.

Reviewer 2: ```suggestion
Note: while the CPU OpenCL back-end supports a sub-group size of 64, Intel's SYCL implementation currently does not. To avoid issues with the sub-group primitives always consider the sub-group size of 64 as not supported by the device.
```

Reviewer: ```diff
-To avoid issues with the sub-group primitives always consider the sub-group size of 64 as not supported by the device.
+To avoid issues with the sub-group primitives, alpaka always considers the sub-group size of 64 as not supported by the device.
```

  ```
  See [Intel's FAQ](https://github.com/intel/compute-runtime/blob/master/opencl/doc/FAQ.md#feature-double-precision-emulation-fp64) for more information.
* The FPGA back-end does not support atomics. alpaka will not check this.
* Device global variables (corresponding to `__device__` and `__constant__` variables in CUDA) are not supported in the SYCL back-end yet.
* Shared memory does not yet work properly on CPU back-ends. The tests that use it (`atomicTest`, `blockSharedTest` and `blockSharedSharingTest`) fail with a `PI_ERROR_OUT_OF_RESOURCES`.

### Choosing the sub-group size (warp size)

Most SYCL targets support multiple sub-group sizes. There is a trait to specify at compile time the sub-group size to use for a kernel that can be defined as follow:

> ```cpp
> template<std::uint32_t TWarpSize, typename TAcc>
> struct alpaka::trait::WarpSize<MyKernel<TWarpSize>, TAcc>
>     : std::integral_constant<std::uint32_t, TWarpSize>
> {
> };
> ```
> `MyKernel` must therefore be templated on the sub-group size.

Reviewer: ```suggestion
For example, if `MyKernel` requires a sub-group size of `32`, this can be declared specialising the `alpaka::trait::WarpSize`:
```cpp
struct MyKernel { ... };

template<typename TAcc>
struct alpaka::trait::WarpSize<MyKernel, TAcc>
    : std::integral_constant<std::uint32_t, 32>
{
};
```
This can be extended to kernels that support multiple sub-group sizes at compile time:
```cpp
template<std::uint32_t TWarpSize>
struct MyKernel { ... };

template<std::uint32_t TWarpSize, typename TAcc>
struct alpaka::trait::WarpSize<MyKernel<TWarpSize>, TAcc>
    : std::integral_constant<std::uint32_t, TWarpSize>
{
};
```
```
